### PR TITLE
Implement #123 - Validate currency code

### PR DESCRIPTION
**Pull request by lionel-nj**
**Summary:**

This PR suggests code to validate currency codes from parsed entities.

**Expected behavior:** 

The validity of field currency_type from fare_attributes.txt will be verified before operations on  parsed entities representing row from fare_attributes.txt.

Capture d’écran, le 2020-04-28 à 17 11 32

Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Run the unit tests with `gradle test` to make sure you didn't break anything
- [x] Format the title like "Fix #<issue_number> - <short description of fix and changes>" (for example - "Fix #1111 - Check for null value before using field")
- [x] Linked all relevant issues
- [x] Include screenshot(s) showing how this pull request works and fixes the issue(s)

### CODE REVIEW

File: adapter/validator/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsTypeValidatorTest.java [hunk 1]
```diff
@@ -1141,4 +1141,43 @@ void emptyRequiredValueGenerateError() {
...
+                mock(IntegerValidator.class),
+                mock(UrlValidator.class),
+                mock(RegexValidator.class),
+                mock(RegexValidator.class),
+                Collections.emptySet()
+        );
+
+        Collection<Notice> result = underTest.validate(new ParsedEntity(
+                TEST_ID,
+                Map.of("currency_type", "JAN"), //
```
Comment (barbeau) [inline adapter/validator/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsTypeValidatorTest.java:1169]: For all rules/notices, we should be checking for potential false positives too. So, we should test against a valid currency type and make sure that no notices are generated.

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/base/ErrorNotice.java [hunk 2]
```diff
@@ -35,6 +35,7 @@
     protected static final String E_015 = "E015";
     protected static final String E_016 = "E016";
     protected static final String E_017 = "E017";
+    protected static final String E_018 = "E018";
```
Comment (barbeau) [inline domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/base/ErrorNotice.java:38]: I would strongly suggest that we start documenting each of these rules in a human readable markdown file like https://github.com/CUTR-at-USF/gtfs-realtime-validator/blob/master/RULES.md along with each of these PRs.

From my experience with the RT validator, it's far easier to document these as we go, as opposed to trying to go back later and re-read the code to remember what we implemented.

File: adapter/validator/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsTypeValidatorTest.java (hunk 1 above)
Comment (fabrice-v) [reply adapter/validator/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsTypeValidatorTest.java:1169]: I agree. Our tests should always include the regular situation where everything goes according to plan.

File: domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/base/ErrorNotice.java (hunk 2 above)
Comment (lionel-nj) [reply domain/src/main/java/org/mobilitydata/gtfsvalidator/domain/entity/notice/base/ErrorNotice.java:38]: Opening an issue on this matter

**Comment by fabrice-v** [comment]
@barbeau While I do agree with your points, I think here we are in a bit of a special case. The goal of the architecture is to have one rule = one use case = one test class.
Here is a special case as we are not following this general idea but we do have a `ValidateGtfsTypes` use case class. It does lacks a test class, and I think fixing that could satisfy the need for readability.

File: adapter/validator/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsTypeValidatorTest.java [hunk 3]
```diff
@@ -1141,4 +1141,51 @@ void emptyRequiredValueGenerateError() {
         assertEquals("Missing value for field:type_date marked as required in entity with id:test_id",
                 notice.getDescription());
     }
+
+    @Test
+    void checkCurrencyCodeValidation() {
```
Comment (fabrice-v) [inline adapter/validator/src/test/java/org/mobilitydata/gtfsvalidator/validator/GtfsTypeValidatorTest.java:1146]: After discussion, not sure we gain in clarity. Please refactor to seperate the two tests and give them specific methods names

**Pull request merged**